### PR TITLE
Ai/ctp 140 create a behaviour tree

### PR DESCRIPTION
**Pull request by ollieiscool**
Implemented very basic chasing of the player along with the spine of the structure of the behaviour tree.

Next steps will be to add patrolling.

### CODE REVIEW

File: CSC8503CoreClasses/GuardObject.h [hunk 1]
```diff
@@ -28,6 +30,21 @@ namespace NCL {
             GameObject* mSightedObject;
             GameObject* mPlayer;
             GameWorld* mWorld;
+        private:
+            bool CanSeePlayer;
+            bool HasCaughtPlayer;
+            bool HasConfiscatedItems;
```
Comment (Nijelous) [inline CSC8503CoreClasses/GuardObject.h:36]: mCanSeePlayer
mHasCaughtPlayer
mHasConfiscatedItems

File: CSC8503CoreClasses/GuardObject.cpp [hunk 2]
```diff
@@ -2,42 +2,175 @@
...
+				if (dist < 5) {
+					HasCaughtPlayer == true;
+					return Failure;
+				}
+			}
+			else if (CanSeePlayer == false && HasCaughtPlayer == false) {
+				return Success;
+			}
+			else if (CanSeePlayer == true && HasCaughtPlayer == true) {
+				return Failure;
```
Comment (Nijelous) [inline CSC8503CoreClasses/GuardObject.cpp:125]: if this returns Failure anyway is there any point in this check? why not just let it fall through to else?

**Review by Nijelous — CHANGES_REQUESTED** [review]
Couple of changes but overall looks good

File: CSC8503CoreClasses/GuardObject.cpp [hunk 3]
```diff
@@ -2,42 +2,175 @@
...
 void GuardObject::RaycastToPlayer() {
 	mPlayer = dynamic_cast<PlayerObject*>(mPlayer);
 	Vector3 dir = (mPlayer->GetTransform().GetPosition() - this->GetTransform().GetPosition()).Normalised();
-	RayCollision closestCollision;
-	Ray r = Ray(this->GetTransform().GetPosition(), dir);
 	float ang = Vector3::Dot(dir, AngleOfSight());
 	if (ang > 2) {
+		RayCollision closestCollision;
+		Ray r = Ray(this->GetTransform().GetPosition(), dir);
 		if (mWorld->Raycast(r, closestCollision, true, this)) {
```
Comment (KyriakosNU) [inline CSC8503CoreClasses/GuardObject.cpp:39]: Is there a way to reduce if statements here?
What if we combined the one in ln 36 and 39 to 
if (ang>2 && mWorld->Raycast(r, closestCollision, true, this))?

**Review by KyriakosNU — APPROVED** [review]

File: CSC8503CoreClasses/GuardObject.h (hunk 1 above)
Comment (ollieiscool) [reply CSC8503CoreClasses/GuardObject.h:36]: Sound arrkid

File: CSC8503CoreClasses/GuardObject.cpp (hunk 3 above)
Comment (ollieiscool) [reply CSC8503CoreClasses/GuardObject.cpp:39]: I split it into two so the raycast didn't run if the player wasn't in front of the guard. So it wont be raycasting constantly even when the player isn't seen.

**Review by Nijelous — APPROVED** [review]

**Pull request merged**